grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/HttpStreamReader.java

```diff
@@ -20,202 +20,105 @@
 import io.grpc.Status;
+import io.netty.buffer.ByteBufAllocator;
 
 /**
- * A {@link Subscriber} to read HTTP messages and pass to gRPC business logic.
+ * A {@link Processor} to read HTTP messages and pass to gRPC business logic.
  */
-public final class HttpStreamReader implements Subscriber<HttpObject> {
+public final class HttpStreamReader extends ArmeriaMessageDeframer {
 
     private final DecompressorRegistry decompressorRegistry;
     private final TransportStatusListener transportStatusListener;
 
-    @VisibleForTesting
-    public final ArmeriaMessageDeframer deframer;
-
-    @Nullable
-    private Subscription subscription;
-
-    private int deferredInitialMessageRequest;
-
-    private volatile boolean cancelled;
-
-    private boolean sawLeadingHeaders;
-
     public HttpStreamReader(DecompressorRegistry decompressorRegistry,
-                            ArmeriaMessageDeframer deframer,
-                            TransportStatusListener transportStatusListener) {
+                            TransportStatusListener transportStatusListener,
+                            ByteBufAllocator alloc, int maxMessageSizeBytes, boolean decodeBase64) {
+        super(alloc, maxMessageSizeBytes, decodeBase64);
         this.decompressorRegistry = requireNonNull(decompressorRegistry, "decompressorRegistry");
-        this.deframer = requireNonNull(deframer, "deframer");
         this.transportStatusListener = requireNonNull(transportStatusListener, "transportStatusListener");
     }
 
-    // Must be called from the IO thread.
-    public void request(int numMessages) {
-        if (subscription == null) {
-            deferredInitialMessageRequest += numMessages;
-            return;
-        }
-        deframer.request(numMessages);
-        requestHttpFrame();
-    }
-
-    // Must be called from the IO thread.
     @Override
-    public void onSubscribe(Subscription subscription) {
-        this.subscription = subscription;
-        if (cancelled) {
-            subscription.cancel();
+    protected void processHeaders(HttpHeaders headers, HttpDeframerOutput<DeframedMessage> out) {
+        // Only clients will see headers from a stream. It doesn't hurt to share this logic between server
+        // and client though as everything else is identical.
+        final String statusText = headers.get(HttpHeaderNames.STATUS);
+        if (statusText == null) {
+            // Not allowed to have empty leading headers, kill the stream hard.
+            transportStatusListener.transportReportStatus(
+                    Status.INTERNAL.withDescription("Missing HTTP status code"));
             return;
         }
-        if (deferredInitialMessageRequest > 0) {
-            request(deferredInitialMessageRequest);
-        }
-    }
 
-    @Override
-    public void onNext(HttpObject obj) {
-        if (cancelled) {
+        final HttpStatus status = HttpStatus.valueOf(statusText);
+        if (!status.equals(HttpStatus.OK)) {
```

Comment:
Shouldn't we handle informational headers?

Reply:
Informational headers is not handled by `processHeaders()`.
It is handled by `processInformationalHeaders()`. `HttpStreamReader` does not override the default behavior that ignores informational headers.